### ISSUE TITLE
error when loading an external logger plugin alongside a builtin one

### ISSUE THREAD

**Issue by doy-stripe** [facebook/osquery#4152]
tl;dr: when using `--logger_plugin=filesystem,test-logger` (where `test-logger` is the name of a plugin loaded from an external extension), the extension manager dies with `terminate called after throwing an instance of 'std::out_of_range'  what():  map::at` after loading the extension.

I'm working on developing a custom logger extension, and I'm running into a bug where loading the logger plugin works fine when it is the only logger plugin loaded, but causes osquery to throw an error when it is loaded with any of the builtin loggers. [Here](https://gist.github.com/doy-stripe/e0504f6e75f348321eca7f2f78071d33) is the source to an extension that reproduces this issue. It's using osquery-go rather than the C++ SDK - I don't *think* that should matter, but it might, I guess.

When I run `sudo osqueryd --extensions_autoload=extensions.load --logger_plugin=test-logger --verbose`, I get this (successful) output:

```
I0223 21:32:53.529295 86395 init.cpp:368] osquery initialized [version=2.11.0]
I0223 21:32:53.543350 86395 system.cpp:344] Found stale process for osqueryd (84082)
I0223 21:32:53.543396 86395 system.cpp:377] Writing osqueryd pid (86395) to /var/run/osqueryd.pidfile
I0223 21:32:53.543459 86395 extensions.cpp:346] Found autoloadable extension: /usr/local/osquery_extensions/test.ext
I0223 21:32:53.543933 86401 watcher.cpp:549] osqueryd watcher (86395) executing worker (86402)
I0223 21:32:53.544157 86401 watcher.cpp:591] Created and monitoring extension child (86403): /usr/local/osquery_extensions/test.ext
I0223 21:32:53.551705 86402 init.cpp:365] osquery worker initialized [watcher=86395]
I0223 21:32:53.551975 86402 rocksdb.cpp:132] Opening RocksDB handle: /var/osquery/osquery.db
2018/02/23 21:32:53 initializing test-plugins
2018/02/23 21:32:53 waiting for socket /var/osquery/osquery.em to exist...
I0223 21:32:53.577345 86415 interface.cpp:338] Extension manager service starting: /var/osquery/osquery.em
2018/02/23 21:32:54 registering plugins
2018/02/23 21:32:54 running extension
I0223 21:32:54.553108 86455 interface.cpp:141] Registering extension (test-plugins, 1908, version=, sdk=)
I0223 21:32:54.553154 86455 registry.cpp:351] Extension 1908 registered logger plugin test-logger
2018/02/23 21:32:54 init: osqueryd
2018/02/23 21:32:54 status: {"s":0,"f":"init.cpp","i":365,"m":"osquery worker initialized [watcher=86395]","h":"[REDACTED]","c":"Fri Feb 23 21:32:53 2018 UTC","u":1519421573}
2018/02/23 21:32:54 status: {"s":0,"f":"rocksdb.cpp","i":132,"m":"Opening RocksDB handle: /var/osquery/osquery.db","h":"[REDACTED]","c":"Fri Feb 23 21:32:53 2018 UTC","u":1519421573}
2018/02/23 21:32:54 status: {"s":0,"f":"interface.cpp","i":338,"m":"Extension manager service starting: /var/osquery/osquery.em","h":"[REDACTED]","c":"Fri Feb 23 21:32:53 2018 UTC","u":1519421573}
2018/02/23 21:32:54 status: {"s":0,"f":"interface.cpp","i":141,"m":"Registering extension (test-plugins, 1908, version=, sdk=)","h":"[REDACTED]","c":"Fri Feb 23 21:32:54 2018 UTC","u":1519421574}
2018/02/23 21:32:54 status: {"s":0,"f":"registry.cpp","i":351,"m":"Extension 1908 registered logger plugin test-logger","h":"[REDACTED]","c":"Fri Feb 23 21:32:54 2018 UTC","u":1519421574}
I0223 21:32:54.557971 86402 events.cpp:825] Event publisher not enabled: audit: Publisher disabled via configuration
I0223 21:32:54.558014 86402 events.cpp:825] Event publisher not enabled: syslog: Publisher disabled via configuration
I0223 21:32:54.558099 86402 events.cpp:1078] Error registering subscriber: socket_events: Subscriber disabled via configuration
I0223 21:32:54.558203 86458 events.cpp:746] Starting event publisher run loop: inotify
I0223 21:32:54.558212 86402 main.cpp:106] Not starting the distributed query service: Distributed query service not enabled.
I0223 21:32:54.558244 86459 events.cpp:746] Starting event publisher run loop: udev
2018/02/23 21:32:54 status: {"s":0,"f":"events.cpp","i":825,"m":"Event publisher not enabled: audit: Publisher disabled via configuration","h":"[REDACTED]","c":"Fri Feb 23 21:32:54 2018 UTC","u":1519421574}
2018/02/23 21:32:54 status: {"s":0,"f":"events.cpp","i":825,"m":"Event publisher not enabled: syslog: Publisher disabled via configuration","h":"[REDACTED]","c":"Fri Feb 23 21:32:54 2018 UTC","u":1519421574}
2018/02/23 21:32:54 status: {"s":0,"f":"events.cpp","i":1078,"m":"Error registering subscriber: socket_events: Subscriber disabled via configuration","h":"[REDACTED]","c":"Fri Feb 23 21:32:54 2018 UTC","u":1519421574}
2018/02/23 21:32:54 status: {"s":0,"f":"events.cpp","i":746,"m":"Starting event publisher run loop: inotify","h":"[REDACTED]","c":"Fri Feb 23 21:32:54 2018 UTC","u":1519421574}
2018/02/23 21:32:54 status: {"s":0,"f":"main.cpp","i":106,"m":"Not starting the distributed query service: Distributed query service not enabled.","h":"[REDACTED]","c":"Fri Feb 23 21:32:54 2018 UTC","u":1519421574}
2018/02/23 21:32:54 status: {"s":0,"f":"events.cpp","i":746,"m":"Starting event publisher run loop: udev","h":"[REDACTED]","c":"Fri Feb 23 21:32:54 2018 UTC","u":1519421574}
```

But if I instead run `sudo osqueryd --extensions_autoload=extensions.load --logger_plugin=filesystem,test-logger --verbose`, this happens:

```
I0223 21:31:48.202580 84082 init.cpp:368] osquery initialized [version=2.11.0]
I0223 21:31:48.213490 84082 system.cpp:344] Found stale process for osqueryd (83514)
I0223 21:31:48.213531 84082 system.cpp:377] Writing osqueryd pid (84082) to /var/run/osqueryd.pidfile
I0223 21:31:48.213593 84082 extensions.cpp:346] Found autoloadable extension: /usr/local/osquery_extensions/test.ext
I0223 21:31:48.214121 84083 watcher.cpp:549] osqueryd watcher (84082) executing worker (84084)
I0223 21:31:48.214355 84083 watcher.cpp:591] Created and monitoring extension child (84085): /usr/local/osquery_extensions/test.ext
2018/02/23 21:31:48 initializing test-plugins
2018/02/23 21:31:48 waiting for socket /var/osquery/osquery.em to exist...
I0223 21:31:48.220993 84084 init.cpp:365] osquery worker initialized [watcher=84082]
I0223 21:31:48.221196 84084 rocksdb.cpp:132] Opening RocksDB handle: /var/osquery/osquery.db
I0223 21:31:48.242049 84096 interface.cpp:338] Extension manager service starting: /var/osquery/osquery.em
2018/02/23 21:31:49 registering plugins
2018/02/23 21:31:49 running extension
I0223 21:31:49.218353 84148 interface.cpp:141] Registering extension (test-plugins, 55813, version=, sdk=)
I0223 21:31:49.218399 84148 registry.cpp:351] Extension 55813 registered logger plugin test-logger
terminate called after throwing an instance of 'std::out_of_range'
  what():  map::at
I0223 21:31:54.215692 84083 watcher.cpp:549] osqueryd watcher (84082) executing worker (84310)
I0223 21:31:54.222234 84310 init.cpp:365] osquery worker initialized [watcher=84082]
I0223 21:31:54.222450 84310 rocksdb.cpp:132] Opening RocksDB handle: /var/osquery/osquery.db
I0223 21:31:54.242370 84317 interface.cpp:338] Extension manager service starting: /var/osquery/osquery.em
E0223 21:31:57.252949 84310 init.cpp:545] Cannot activate filesystem,test-logger logger plugin: Unknown registry plugin: test-logger
```

It appears that something is causing the extension manager to crash, and when it gets restarted, it doesn't have state about the extensions that were previously loaded, and so it then fails to load the logger plugin from the extension.

**Comment by theopolis** [comment]
Definitely a bug!

```
» lldb ../osquery/build/darwin/osquery/osqueryd 
(lldb) target create "../osquery/build/darwin/osquery/osqueryd"
Current executable set to '../osquery/build/darwin/osquery/osqueryd' (x86_64).
(lldb) run --ephemeral --logger_plugin=foobar_logger,filesystem --extension ./examples/foobar_logger.ext --verbose --extensions_autoload /dev/null --extensions_socket /tmp/em.sock --disable_watchdog
Process 80789 launched: '../osquery/build/darwin/osquery/osqueryd' (x86_64)
I0224 18:46:09.857192 2393895744 init.cpp:388] osquery initialized [version=3.1.0-29-g3616dcc4]
2018-02-24 18:46:09.859025-0800 osqueryd[80789:8770623] osqueryd started [version=3.1.0-29-g3616dcc4]
I0224 18:46:09.859525 236994560 interface.cpp:381] Extension manager service starting: /tmp/em.sock
I0224 18:46:09.860119 235384832 watcher.cpp:591] Created and monitoring extension child (80796): ./examples/foobar_logger.ext
W0224 18:46:09.860150 2393895744 init.cpp:651] Error reading config: config file does not exist: /var/osquery/osquery.conf
I0224 18:46:10.101331 238055424 interface.cpp:149] Registering extension (foobar_logger, 24760, version=1.0.0, sdk=1.8.0)
I0224 18:46:10.101807 238055424 registry.cpp:351] Extension 24760 registered logger plugin foobar_logger
libc++abi.dylib: terminating with uncaught exception of type std::out_of_range: map::at:  key not found
Process 80789 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x00007fff55bf5e3e libsystem_kernel.dylib`__pthread_kill + 10
libsystem_kernel.dylib`__pthread_kill:
->  0x7fff55bf5e3e <+10>: jae    0x7fff55bf5e48            ; <+20>
    0x7fff55bf5e40 <+12>: movq   %rax, %rdi
    0x7fff55bf5e43 <+15>: jmp    0x7fff55bed0b8            ; cerror_nocancel
    0x7fff55bf5e48 <+20>: retq   
Target 0: (osqueryd) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
  * frame #0: 0x00007fff55bf5e3e libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff55d34150 libsystem_pthread.dylib`pthread_kill + 333
    frame #2: 0x00007fff55b52312 libsystem_c.dylib`abort + 127
    frame #3: 0x00007fff53b2df8f libc++abi.dylib`abort_message + 245
    frame #4: 0x00007fff53b2e113 libc++abi.dylib`default_terminate_handler() + 241
    frame #5: 0x00007fff54eb8eab libobjc.A.dylib`_objc_terminate() + 105
    frame #6: 0x00007fff53b497c9 libc++abi.dylib`std::__terminate(void (*)()) + 8
    frame #7: 0x00007fff53b4926d libc++abi.dylib`__cxa_throw + 121
    frame #8: 0x0000000100616917 osqueryd`std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, unsigned long long, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, unsigned long long> > >::at(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 139
    frame #9: 0x00000001006165e5 osqueryd`osquery::RegistryInterface::setActive(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 483
    frame #10: 0x0000000100618f65 osqueryd`osquery::RegistryFactory::setActive(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 91
    frame #11: 0x00000001005c344f osqueryd`std::__1::__function::__func<osquery::Initializer::initActivePlugin(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const::$_0, std::__1::allocator<osquery::Initializer::initActivePlugin(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const::$_0>, osquery::Status (bool&)>::operator()(bool&) + 57
    frame #12: 0x00000001005f711d osqueryd`std::__1::function<osquery::Status (bool&)>::operator()(bool&) const + 41
    frame #13: 0x00000001005f7086 osqueryd`osquery::applyExtensionDelay(std::__1::function<osquery::Status (bool&)>) + 147
    frame #14: 0x00000001005c2462 osqueryd`osquery::Initializer::initActivePlugin(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const + 150
    frame #15: 0x00000001005c2c0e osqueryd`osquery::Initializer::start() const + 1268
    frame #16: 0x0000000100513cec osqueryd`osquery::startOsquery(int, char**, std::__1::function<void ()>) + 418
    frame #17: 0x000000010051710b osqueryd`main + 37
    frame #18: 0x00007fff55aa6115 libdyld.dylib`start + 1
    frame #19: 0x00007fff55aa6115 libdyld.dylib`start + 1
```
